### PR TITLE
Detect template loop formed in Lua code

### PR DESCRIPTION
**Pull request by xxyzz**
Fix tatuylonen/wiktextract#894

`test_template16` is removed because it's not a valid test, it has infinite loop.

I can't reproduce the loop in test, but it has the same repeated expand stack as the error message. The test stops at "too deep recursion" error without the change but it fixes the loop in page "今生".

### CODE REVIEW

File: src/wikitextprocessor/core.py [hunk 1]
```diff
@@ -1909,3 +1918,14 @@ def create_strip_marker(self, node: str, content: str) -> str:
                 self.strip_marker_cache["preprocess"] += 1
                 self.strip_marker_cache[content] = num
         return f"""\x7f'"`UNIQ--{node}-{num_str}-QINU`"'\x7f"""
+
+
+def detect_expand_template_loop(stack: list[str]) -> bool:
+    stack_len = len(stack)
```
Comment (kristian-clausal) [inline src/wikitextprocessor/core.py:1924]: Do you think we could add a couple of guard clauses here? Like `if stack[-1] not in stack[:-1]: return False`? We could also return True when stack_len is small, and only start checking when it grows too big.

File: src/wikitextprocessor/core.py [hunk 2]
```diff
@@ -1410,7 +1410,14 @@ def expander(arg: str) -> str:
                         continue
 
                     # Construct and expand template arguments
-                    self.expand_stack.append(name)
+                    self.expand_stack.append("Template:" + name)
+                    if detect_expand_template_loop(self.expand_stack):
+                        parts.append(
```
Comment (kristian-clausal) [inline src/wikitextprocessor/core.py:1415]: Should we have a `wxr.wtp.error` or `wxr.wtp.warning` here? We are handling the issue, and it's caused by bad code on Wiktionary's side, so not necessarily.

**Review by kristian-clausal — APPROVED** [review]

**Comment by kristian-clausal** [comment]
I had only minor comments that I'm not sure are valid, otherwise this looks good. I'm worried how performant the loop checking is, but that's not something we can avoid easily.

**Comment by xxyzz** [comment]
Add `mw.logObject(template:get_arguments())` at line 93 in https://en.wiktionary.org/wiki/Module:zh-translit and preview page "今生", the Lua log will show loop detected for template "m".

**Comment by kristian-clausal** [comment]
Do you think they know about this on the Wiktionary side of things?

**Comment by xxyzz** [comment]
It's not a problem on Wiktionary because the parser could stop the loop, I think maybe Lua module editors don't need to worry about it.

File: src/wikitextprocessor/core.py (hunk 1 above)
Comment (xxyzz) [reply src/wikitextprocessor/core.py:1924]: Short list also could have loop and check it is fast. I'll add the "if" check.

File: src/wikitextprocessor/core.py (hunk 2 above)
Comment (xxyzz) [reply src/wikitextprocessor/core.py:1415]: I'll add a warning here.

File: src/wikitextprocessor/core.py (hunk 1 above)
Comment (xxyzz) [reply src/wikitextprocessor/core.py:1924]: Maybe there is a better algorithm for this function, I'm also not sure how this would impact performance.

Comment (kristian-clausal) [reply src/wikitextprocessor/core.py:1928]: We can merge now that it's working, see how much performance is affected and try to figure out something later.

**Comment by xxyzz** [comment]
There is a discussion about the bug on Wiktionary: https://en.wiktionary.org/wiki/User_talk:Theknightwho#Pages_with_template_loops

**Pull request merged**